prover_server/prover_server.cpp

```diff
@@ -520,34 +455,49 @@ int main(int argc, char **argv)
                               &prover]() {
         if (boost::filesystem::exists(keypair_file)) {
             std::cout << "[INFO] Loading keypair: " << keypair_file << "\n";
-            return load_keypair(keypair_file);
+
+            snark::keypair keypair;
+            std::ifstream in_s =
+                libtool::open_binary_input_file(keypair_file.c_str());
+            snark::keypair_read_bytes(keypair, in_s);
+            return keypair;
         }
 
         std::cout << "[INFO] No keypair file " << keypair_file
                   << ". Generating.\n";
         const snark::keypair keypair = prover.generate_trusted_setup();
-        std::cout << "[INFO] Writing new keypair to " << keypair_file << "\n";
-        write_keypair(keypair, keypair_file);
 
         if (!proving_key_output_file.empty()) {
             std::cout << "[DEBUG] Writing separate proving key to "
                       << proving_key_output_file << "\n";
-            write_proving_key(keypair.pk, proving_key_output_file);
+            std::ofstream out_s = libtool::open_binary_output_file(
+                proving_key_output_file.c_str());
+            snark::proving_key_write_bytes(keypair.pk, out_s);
         }
         if (!verification_key_output_file.empty()) {
             std::cout << "[DEBUG] Writing separate verification key to "
                       << verification_key_output_file << "\n";
-            write_verification_key(keypair.vk, verification_key_output_file);
+            std::ofstream out_s = libtool::open_binary_output_file(
+                verification_key_output_file.c_str());
+            snark::verification_key_write_bytes(keypair.vk, out_s);
         }
 
+        // Write the keypair last. If something above fails, this same
+        // code-path will be executed again on the next invocation.
```

Comment:
Just for my understanding what's the worry here if the keypair is written first?
Do you fear that in the event of an error a keypair could fail to be written on file entirely but the corresponding pk could have already been written leading to some potential inconsistencies in what's written to files?

Reply:
If the keypair is written first, but something fails afterwards, we would have to manually remove the keypair and re-run in order to re-try the other operations (since the control flow depends on the existence of the keypair).